src/celeritas/optical/model/MieModel.cc

```diff
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/optical/model/MieModel.cc
+//---------------------------------------------------------------------------//
+#include "MieModel.hh"
+
+#include "corecel/Assert.hh"
+#include "corecel/Types.hh"
+#include "corecel/data/CollectionBuilder.hh"
+#include "corecel/io/Logger.hh"
+#include "celeritas/Types.hh"
+#include "celeritas/inp/Grid.hh"
+#include "celeritas/optical/CoreParams.hh"
+#include "celeritas/optical/CoreState.hh"
+#include "celeritas/optical/InteractionApplier.hh"
+#include "celeritas/optical/MfpBuilder.hh"
+#include "celeritas/optical/MieData.hh"
+#include "celeritas/optical/action/ActionLauncher.hh"
+#include "celeritas/optical/action/TrackSlotExecutor.hh"
+#include "celeritas/optical/model/MieExecutor.hh"
+
+namespace celeritas
+{
+namespace optical
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Create a model builder from imported data.
+ */
+auto MieModel::make_builder(SPConstImported imported, Input input)
+    -> ModelBuilder
+{
+    CELER_EXPECT(imported);
+    return [imported = std::move(imported),
+            input = std::move(input)](ActionId id) {
+        return std::make_shared<MieModel>(id, imported, input);
+    };
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct the model from imported data and imported material parameters.
+ */
+MieModel::MieModel(ActionId id, SPConstImported imported, Input input)
+    : Model(id, "optical-mie", "interact by optical Mie scattering")
+    , imported_(ImportModelClass::mie, std::move(imported))
+{
+    HostVal<MieData> data;
+    CollectionBuilder builder{&data.mie_record};
+
+    for (auto const& mie : input.data)
+    {
+        MieMaterialData record;
+        record.backward_g = mie.backward_g;
+        record.forward_g = mie.forward_g;
+        record.forward_ratio = mie.forward_ratio;
+
+        builder.push_back(record);
+    }
+
+    data_ = CollectionMirror<MieData>{std::move(data)};
+    CELER_ENSURE(data_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Build the mean free paths for the model.
+ */
+void MieModel::build_mfps(OptMatId mat, MfpBuilder& build) const
+{
+    CELER_EXPECT(mat < imported_.num_materials());
+
+    if (auto const& mfp = imported_.mfp(mat))
+    {
+        build(mfp);
+    }
+    else
+    {
+        CELER_LOG(debug)
+            << "mie model MFP not found for setting to infinity for "
+            << mat.get();
+        inp::Grid g;
+        // should change these at some point. the .x vals are coming from the
+        // opnovice example as of now
+        g.x = {1.56962, 6.19998};
+        g.y = {std::numeric_limits<real_type>::infinity(),
+               std::numeric_limits<real_type>::infinity()};
+        build(g);
```

Comment:
I assume the energy bounds are stray debugging? I think this would work:
```suggestion
        build(inp::Grid::from_constant(std::numeric_limits<double>::infinity()));
```

(or should we change the MfpBuilder to construct an infinite MFP over all energies instead of an invalid grid?)

Reply:
It is for debugging purpose. I thought it will be good to handle the `else` condition but was not sure of a cleaner way to get the `x` values- hence hardcoded.

Reply:
I agree, just like the other models where there's "incomplete" cross sections. I think as long as `x` encompasses the energy range of the optical photons we care about, we'll be ok: hence the "constant" (which uses `[0, inf)`).